### PR TITLE
Creates a common 'addIndex' function.  Closes #484.

### LINKED ISSUE #484: Replace all the .idx functions with a common addIdx
I'm far from certain about this, but based on some discussions recently in #452.  (It's a long thread, you can start with comments from [2014-10-27][ad]) I was thinking about replacing all the controversial `xxx.idx` functions with a single wrapper for callback functions:

```javascript
    /**
     * Creates a wrapper for an iteration callback, such as one that might be
     * supplied to `map`, or `reduce`.  This wrapper supplies the index as well
     * as the standard parameters usually passed.  If the optional `list`
     * parameter is supplied to this function, it also passes that into each
     * iteration.
     *
     * @func
     * @memberOf R
     * @category Function
     * @sig (a -> b -> ... -> n) -> (a -> b -> ... Num -> n)
     * @sig (a -> b -> ... -> n) -> (a -> b -> ... Num -> {*} -> n)
     * @param {Function} fn A list iteration callback function.
     * @param {List} [list] The list that will be iterated
     * @return {Function} A new function that supplies the index (and optionally
     *         the entire list) as well as the existing parameters.
     * @example
     *
     *      var primes = [2, 3, 5, 7, 11];
     *
     *      var convert1 = function(prime, idx) {return prime + "-" + idx;};
     *      map(addIdx(convert1), primes);
     *      //=> ["2-0", "3-1", "5-2", "7-3", "11-4"]
     *
     *      var convert2 = function(prime, idx, list) {
     *          return prime + "-" + idx + "/" + list.length;
     *      };
     *      map(addIdx(convert2, primes), primes)
     *      //=> ["2-0/5", "3-1/5", "5-2/5", "7-3/5", "11-4/5"]
     *
     *      var convert3 = function(soFar, prime, idx) {
     *          return soFar + prime / (idx + 1);
     *      };
     *      reduce(addIdx(convert3), 0, primes);
     *      //=> 2/1 + 3/2 + 5/3 + 7/4 + 11/5 = 2735/60 ~= 9.116666666666667
     */
    R.addIdx = function(fn, list) {
        var idx = 0;
        return function() {
            return fn.apply(this, concat(
                _slice(arguments),
                list ? [idx++, list] : idx++
            ));
        }
    };
```

I haven't tested deeply, but it works for the examples in the comments.  The nicest bit is that most of the time people want these, they just want the index and not the list, and then the wrapper is fairly simple, just a wrap around their function reference or their lambda.

The downside is that it's easy enough to see -- even without testing -- that this will be slower than the current technique.

Comments?

  [ad]: https://github.com/ramda/ramda/issues/452#issuecomment-60602467

### PR DESCRIPTION
**Pull request by CrossEye**
This has been a long time coming.  It addresses only list function, not object ones.

### CODE REVIEW

File: src/addIndex.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+var curryN = require('./curryN');
+
+
+/**
+ * Creates a new list iteration function from an existing one by adding two new parameters
+ * to its callback function: the current index, and the entire list.
+ *
+ * This would turn, for instance, Ramda's simple `map` function into one that more closely
+ * resembles `Array.prototype.map`.  Note that this will only work for functions in which
+ * the iteration callback function is first the first parameter, and where the list is the
```
Comment (davidchambers) [inline src/addIndex.js]: s/is first the first parameter/is the first parameter/

File: src/addIndex.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ * @memberOf R
+ * @category Function
+ * @category List
+ * @sig ((a -> b) -> ... -> [a] -> *) -> (a -> Int -> [a] -> b) -> ... -> [a] -> *)
+ * @param {Function} fn A list iteration function that does not pass index/list to its callback
+ * @return An altered list iteration function thats passes index/list to its callback
+ * @example
+ *
+ *      var mapIndexed = R.addIndex(R.map);
+ *      mapIndexed(function(val, idx) {return idx + '-" + val;}, ['f', 'o', 'o', 'b', 'a', 'r']);
```
Comment (davidchambers) [inline src/addIndex.js]: s/"/'/

File: src/mapIndexed.js [hunk 3]
```diff
@@ -16,6 +16,8 @@ var _curry2 = require('./internal/_curry2');
  * @param {Function} fn The function to be called on every element of the input `list`.
  * @param {Array} list The list to be iterated over.
  * @return {Array} The new list.
+ * @deprecated since v0.14.0
+ * @see R.addIndex
```
Comment (davidchambers) [inline src/mapIndexed.js]: :grinning:

File: test/addIndex.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+      return (idx === 0 || idx === list.length - 1) ? x * x : x;
+    };
+    var mapIndexed = R.addIndex(R.map);
+
+    it('working just like a normal map', function() {
+      assert.deepEqual(mapIndexed(times2, [1, 2, 3, 4]), [2, 4, 6, 8]);
+    });
+
+    it('passing the index as a second parameter to the callback', function() {
+      assert.deepEqual(mapIndexed(addIndexParam, [8, 6, 7, 5, 3, 0, 9]), [8 + 0, 6 + 1, 7 + 2, 5 + 3, 3 + 4, 0 + 5, 9 + 6]);
```
Comment (davidchambers) [inline test/addIndex.js]: This example would be clearer with a shorter input list.

**Comment by davidchambers** [comment]
This is very exciting to see!

File: test/addIndex.js (hunk 4 above)
Comment (CrossEye) [inline test/addIndex.js]: Yes, but would get [stuck in your head][tt] much less often.

  [tt]: https://www.youtube.com/watch?v=6WTdTwcmxyo

Comment (davidchambers) [inline test/addIndex.js]: I remember having a landline with a seven-digit phone number, and that was only after a digit was added to our six-digit number. :)

File: test/addIndex.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+describe('addIndex', function() {
+  describe('works with unary functions like `map`', function() {
+    var times2 = function(x) {return x * 2;};
+    var addIndexParam = function(x, idx) {return x + idx;};
+    var squareEnds = function(x, idx, list) {
+      return (idx === 0 || idx === list.length - 1) ? x * x : x;
+    };
+    var mapIndexed = R.addIndex(R.map);
+
+    it('working just like a normal map', function() {
```
Comment (buzzdecafe) [inline test/addIndex.js]: caveman test?

File: src/mapIndexed.js (hunk 3 above)
Comment (buzzdecafe) [inline src/mapIndexed.js]: :tada:

File: test/addIndex.js (hunk 5 above)
Comment (CrossEye) [inline test/addIndex.js]: It's hard to decide if it's the source code or the output that we want to look correct.

In other words, ooga ooga.

Comment (buzzdecafe) [inline test/addIndex.js]: :fire: bad

**Comment by davidchambers** [comment]
Squash and merge, I say. :)

**Comment by buzzdecafe** [comment]
ya this is a BFD. well done

**Comment by CrossEye** [comment]
One other question before merging this.  The signature I've given this seems overly complex, but I can't think of anything simpler that's also fairly accurate.  Is there a better version?

**Comment by buzzdecafe** [comment]
` @sig ((a... -> b) -> x? -> [a] -> *) -> ((a..., Int, [a] -> b) -> x? -> [a] -> x)`

I think this is only slightly less ugly. the issue is that this function is doing a lot of magic work to handle `map` and `reduce` at one go.

**Comment by CrossEye** [comment]
I like dropping the initial arrows, and switching to commas is fine.  But I don't buy ` -> x? ->` to represent an unknown number of parameters.  The function would work fine on

    (a -> b) -> c -> d -> e -> [a] -> x

transforming it into

    (a -> Int -> [a] -> b)  -> c -> d -> e -> [a] -> x

I don't feel that `x?` is a good representation of that `c -> d -> e`.

**Comment by wayneseymour** [comment]
I'm sorry I've very little to contribute, but where best to go learn the syntax for your signatures?

**Comment by buzzdecafe** [comment]
@wayneseymour our sigs are like pidgin Haskell sigs. You are better off looking at Haskell (or ML) sigs and then seeing how we are polluting them

**Comment by wayneseymour** [comment]
@buzzdecafe Hahahaha, thanks man!  :)

**Comment by buzzdecafe** [comment]
@wayneseymour you can only graft so much strong functional typing onto js before the cracks begin to show. The sig we're discussing here is a good (bad?) example of that ...

**Comment by wayneseymour** [comment]
@buzzdecafe 
Ah, makes sense. I'll be looking into Haskell's types very soon.

**Comment by CrossEye** [comment]
@wayneseymour: I gave a partial description of it recently: https://github.com/ramda/ramda/issues/1035#issuecomment-95327214

**Comment by davidchambers** [comment]
> I don't feel that `x?` is a good representation of that `c -> d -> e`.

I've been using `?` to mean "or null/undefined". So `String -> String?`, for example, is a function from String to String/null/undefined.

**Comment by wayneseymour** [comment]
@CrossEye
  Ah, thanks

**Comment by buzzdecafe** [comment]
> I've been using ? to mean "or null/undefined". 

That seems reasonable. The problem with the function here is that it really needs *two* sigs, one for `map` type functions, and one for `reduce` type functions, since the `x?` may not be there at all.

**Comment by CrossEye** [comment]
But it's much more generic than that.  It doesn't care _what_ your signature is like as long as the iteration callback comes first and the list comes last.  That's what I want to capture, in any case.

**Comment by buzzdecafe** [comment]
are we just waiting on sig for this? let's merge the code and fix the docs later :smile:

**Comment by CrossEye** [comment]
Somehow I thought I'd already merged this one!

Brainfart.

**Comment by CrossEye** [comment]
I guess I need to update it now, though, since all the deprecation warnings say "since 0.14"...

**Comment by buzzdecafe** [comment]
merging. will fix deprecated notices in a separate pr

**Pull request merged**